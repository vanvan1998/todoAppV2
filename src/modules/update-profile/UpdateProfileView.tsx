'use client';

import React, { useRef, useState } from 'react';
import { isEmpty } from 'lodash';
import styled from 'styled-components';
import { Form, Image } from 'react-bootstrap';
import { useAuth } from 'src/contexts/AuthContext';
import { storage } from '../../firebase';
import { ref, getDownloadURL, uploadBytesResumable } from 'firebase/storage';
import { styles } from './UpdateProfileView.styles';
import { ErrorText, HeaderTitle, SuccessText, Title, Button, Input } from 'src/components';
import { placeholderColor } from 'src/theme';
import { useMediaQuery } from 'src/hooks';

const Container = styled.div`
  ${styles.Container}
`;

const HeaderWrapper = styled.div<{ isMobile: boolean }>`
  ${styles.headerWrapper}
`;

const ErrorWrapper = styled.div<{ isMobile: boolean }>`
  ${styles.errorWrapper}
`;

export const UpdateProfileView = () => {
  const { currentUser, updateProfile } = useAuth();
  const [name, setName] = useState(currentUser?.displayName);
  const [imgUrl, setImgUrl] = useState(currentUser?.photoURL);
  const [error, setError] = useState('');
  const [errorName, setErrorName] = useState('');
  const [success, setSuccess] = useState('');
  const [successName, setSuccessName] = useState('');
  const [loading, setLoading] = useState(false);
  const { isMobile } = useMediaQuery();
  const selectedFileRef = useRef();
  if (!currentUser) return <></>;

  const onFileChange = e => {
    selectedFileRef.current = e.target.files[0];
  };

  const handleSubmitImage = (e: any) => {
    e.preventDefault();

    if (!selectedFileRef.current) return;

    const promises: any[] = [];
    const storageRef = ref(storage, `files/${(selectedFileRef.current as any).name}`);
    const uploadTask = uploadBytesResumable(storageRef, selectedFileRef.current);

    uploadTask.on(
      'state_changed',
      () => {
        // const progress = Math.round(
        //   (snapshot.bytesTransferred / snapshot.totalBytes) * 100
        // );
      },
      error => {
        console.log(error);
      },
      () => {
        getDownloadURL(uploadTask.snapshot.ref).then(downloadURL => {
          setImgUrl(downloadURL);
          promises.push(updateProfile({ photoURL: downloadURL }));
        });
      }
    );

    Promise.all(promises)
      .then(() => {
        setSuccess('Update Avatar successfully');
      })
      .catch(() => {
        setError('Failed to update account');
      })
      .finally(() => {
        setLoading(false);
      });
  };

  const handleSubmit = (e: { preventDefault: () => void }) => {
    e.preventDefault();

    const promises = [];
    setLoading(true);
    setErrorName('');
    setSuccessName('');

    if (name) {
      promises.push(updateProfile({ name }) as never);
    }

    Promise.all(promises)
      .then(() => {
        setSuccessName('Full name updated successfully');
      })
      .catch(() => {
        setErrorName('Failed to update account');
      })
      .finally(() => {
        setLoading(false);
      });
  };

  return (
    <Container>
      <HeaderWrapper isMobile={isMobile}>
        <HeaderTitle style={{ fontWeight: '600' }}>Profile</HeaderTitle>
        <Title>
          You&apos;re logged in as <Title style={{ fontWeight: 'bold' }}>{currentUser.email}</Title>
        </Title>
      </HeaderWrapper>

      <Form style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
        <Form.Group id='file'>
          <div className='mb-4 text-center'>{imgUrl && <Image src={imgUrl} alt='uploaded file' height={300} />}</div>
          <Title style={{ padding: '8px 0' }}>Avatar</Title>
          <Form.Control type='file' onChange={onFileChange} />
        </Form.Group>

        {error && (
          <ErrorWrapper isMobile={isMobile}>
            <ErrorText>{error}</ErrorText>
          </ErrorWrapper>
        )}
        {success && (
          <ErrorWrapper isMobile={isMobile}>
            <SuccessText>{success}</SuccessText>
          </ErrorWrapper>
        )}
        <Button
          disabled={loading}
          title='Update Avatar'
          handleButton={handleSubmitImage}
          styles={{
            marginTop: 8,
            marginBottom: 16,
            ...(isMobile ? { maxWidth: 320, minWidth: 200, width: '80vw' } : { width: 320 })
          }}
          testId='update-avatar-button'
        />
      </Form>
      <Input
        title='Full name'
        value={name}
        onChange={setName}
        styles={{ marginBottom: 8, marginTop: 32 }}
        placeholder='Enter your full name...'
        inputStyles={{
          ...(isMobile ? { maxWidth: 320, minWidth: 200, width: '80vw' } : { minHeight: 48, minWidth: 320 }),
          border: `1px solid ${placeholderColor}`,
          backgroundColor: '#fefeff'
        }}
      />
      {errorName && (
        <ErrorWrapper isMobile={isMobile}>
          <ErrorText>{errorName}</ErrorText>
        </ErrorWrapper>
      )}
      {successName && (
        <ErrorWrapper isMobile={isMobile}>
          <SuccessText>{successName}</SuccessText>
        </ErrorWrapper>
      )}
      <Button
        disabled={loading || isEmpty(name)}
        title='Update'
        handleButton={handleSubmit}
        styles={{
          marginTop: 8,
          marginBottom: 16,
          ...(isMobile ? { maxWidth: 320, minWidth: 200, width: '80vw' } : { width: 320 })
        }}
        testId='update-button'
      />
    </Container>
  );
};
