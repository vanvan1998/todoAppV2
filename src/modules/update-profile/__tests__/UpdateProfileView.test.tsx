import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { UpdateProfileView } from '../UpdateProfileView';

jest.mock('@next/third-parties/google', () => {
  return {
    GoogleAnalytics: () => <>GoogleAnalytics</>
  };
});

jest.mock('src/hooks', () => {
  const originalModule = jest.requireActual('src/hooks');
  return {
    ...originalModule,
    useMediaQuery: jest
      .fn()
      .mockReturnValue({
        isMobile: true
      })
      .mockReturnValueOnce({ isMobile: false })
  };
});

jest.mock('src/contexts', () => {
  return {
    useAuth: jest
      .fn()
      .mockReturnValue({ currentUser: { displayName: 'displayName', email: 'email' }, updateProfile: jest.fn() })
  };
});

jest.mock('next/navigation', () => {
  return {
    useRouter: jest.fn().mockReturnValue({
      push: jest.fn()
    })
  };
});

describe('UpdateProfileView', () => {
  it('renders UpdateProfileView with mobile view', () => {
    const component = render(<UpdateProfileView />);
    expect(component).toMatchSnapshot();
  });

  it('renders UpdateProfileView with desktop view', () => {
    const component = render(<UpdateProfileView />);

    const button = component.getByTestId('update-avatar-button');
    const updateButton = component.getByTestId('update-button');
    fireEvent.click(button);
    fireEvent.click(updateButton);
    expect(component).toMatchSnapshot();
  });
});
