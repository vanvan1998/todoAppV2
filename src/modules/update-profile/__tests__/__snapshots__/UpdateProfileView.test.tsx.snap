// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateProfileView renders UpdateProfileView with desktop view 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dQemvk eeThTR"
      >
        <div
          class="sc-bCDiRt hSyFex"
        >
          <div
            class="sc-beqWNU hnIjJd"
            style="font-weight: 600;"
          >
            Profile
          </div>
          <div
            class="sc-gueXAH gRHGwY"
          >
            You're logged in as 
            <div
              class="sc-gueXAH gRHGwY"
              style="font-weight: bold;"
            >
              email
            </div>
          </div>
        </div>
        <form
          class=""
          style="display: flex; flex-direction: column; align-items: center;"
        >
          <div
            id="file"
          >
            <div
              class="mb-4 text-center"
            />
            <div
              class="sc-gueXAH gRHGwY"
              style="padding: 8px 0px;"
            >
              Avatar
            </div>
            <input
              class="form-control"
              type="file"
            />
          </div>
          <button
            buttontype="primary"
            class="sc-gLDymS dItXMH"
            data-testid="update-avatar-button"
            disabled=""
            style="margin-top: 8px; margin-bottom: 16px; max-width: 320px; min-width: 200px; width: 80vw; background-color: rgb(182, 180, 185);"
            type="button"
          >
            Update Avatar
          </button>
        </form>
        <div
          style="margin-bottom: 8px; margin-top: 32px;"
        >
          <div
            class="sc-gueXAH gRHGwY"
          >
            Full name
          </div>
          <input
            aria-label="Full name"
            class="sc-iAEzLF gTKcTi"
            placeholder="Enter your full name..."
            style="max-width: 320px; min-width: 200px; width: 80vw; border: 1px solid #858585; background-color: rgb(254, 254, 255);"
            type="text"
            value="displayName"
          />
        </div>
        <button
          buttontype="primary"
          class="sc-gLDymS dItXMH"
          data-testid="update-button"
          disabled=""
          style="margin-top: 8px; margin-bottom: 16px; max-width: 320px; min-width: 200px; width: 80vw; background-color: rgb(182, 180, 185);"
          type="button"
        >
          Update
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dQemvk eeThTR"
    >
      <div
        class="sc-bCDiRt hSyFex"
      >
        <div
          class="sc-beqWNU hnIjJd"
          style="font-weight: 600;"
        >
          Profile
        </div>
        <div
          class="sc-gueXAH gRHGwY"
        >
          You're logged in as 
          <div
            class="sc-gueXAH gRHGwY"
            style="font-weight: bold;"
          >
            email
          </div>
        </div>
      </div>
      <form
        class=""
        style="display: flex; flex-direction: column; align-items: center;"
      >
        <div
          id="file"
        >
          <div
            class="mb-4 text-center"
          />
          <div
            class="sc-gueXAH gRHGwY"
            style="padding: 8px 0px;"
          >
            Avatar
          </div>
          <input
            class="form-control"
            type="file"
          />
        </div>
        <button
          buttontype="primary"
          class="sc-gLDymS dItXMH"
          data-testid="update-avatar-button"
          disabled=""
          style="margin-top: 8px; margin-bottom: 16px; max-width: 320px; min-width: 200px; width: 80vw; background-color: rgb(182, 180, 185);"
          type="button"
        >
          Update Avatar
        </button>
      </form>
      <div
        style="margin-bottom: 8px; margin-top: 32px;"
      >
        <div
          class="sc-gueXAH gRHGwY"
        >
          Full name
        </div>
        <input
          aria-label="Full name"
          class="sc-iAEzLF gTKcTi"
          placeholder="Enter your full name..."
          style="max-width: 320px; min-width: 200px; width: 80vw; border: 1px solid #858585; background-color: rgb(254, 254, 255);"
          type="text"
          value="displayName"
        />
      </div>
      <button
        buttontype="primary"
        class="sc-gLDymS dItXMH"
        data-testid="update-button"
        disabled=""
        style="margin-top: 8px; margin-bottom: 16px; max-width: 320px; min-width: 200px; width: 80vw; background-color: rgb(182, 180, 185);"
        type="button"
      >
        Update
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`UpdateProfileView renders UpdateProfileView with mobile view 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dQemvk eeThTR"
      >
        <div
          class="sc-bCDiRt QpqjE"
        >
          <div
            class="sc-beqWNU hnIjJd"
            style="font-weight: 600;"
          >
            Profile
          </div>
          <div
            class="sc-gueXAH gRHGwY"
          >
            You're logged in as 
            <div
              class="sc-gueXAH gRHGwY"
              style="font-weight: bold;"
            >
              email
            </div>
          </div>
        </div>
        <form
          class=""
          style="display: flex; flex-direction: column; align-items: center;"
        >
          <div
            id="file"
          >
            <div
              class="mb-4 text-center"
            />
            <div
              class="sc-gueXAH gRHGwY"
              style="padding: 8px 0px;"
            >
              Avatar
            </div>
            <input
              class="form-control"
              type="file"
            />
          </div>
          <button
            buttontype="primary"
            class="sc-gLDymS dItXMH"
            data-testid="update-avatar-button"
            style="margin-top: 8px; margin-bottom: 16px; width: 320px;"
            type="button"
          >
            Update Avatar
          </button>
        </form>
        <div
          style="margin-bottom: 8px; margin-top: 32px;"
        >
          <div
            class="sc-gueXAH gRHGwY"
          >
            Full name
          </div>
          <input
            aria-label="Full name"
            class="sc-iAEzLF gTKcTi"
            placeholder="Enter your full name..."
            style="min-height: 48px; min-width: 320px; border: 1px solid #858585; background-color: rgb(254, 254, 255);"
            type="text"
            value="displayName"
          />
        </div>
        <button
          buttontype="primary"
          class="sc-gLDymS dItXMH"
          data-testid="update-button"
          style="margin-top: 8px; margin-bottom: 16px; width: 320px;"
          type="button"
        >
          Update
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dQemvk eeThTR"
    >
      <div
        class="sc-bCDiRt QpqjE"
      >
        <div
          class="sc-beqWNU hnIjJd"
          style="font-weight: 600;"
        >
          Profile
        </div>
        <div
          class="sc-gueXAH gRHGwY"
        >
          You're logged in as 
          <div
            class="sc-gueXAH gRHGwY"
            style="font-weight: bold;"
          >
            email
          </div>
        </div>
      </div>
      <form
        class=""
        style="display: flex; flex-direction: column; align-items: center;"
      >
        <div
          id="file"
        >
          <div
            class="mb-4 text-center"
          />
          <div
            class="sc-gueXAH gRHGwY"
            style="padding: 8px 0px;"
          >
            Avatar
          </div>
          <input
            class="form-control"
            type="file"
          />
        </div>
        <button
          buttontype="primary"
          class="sc-gLDymS dItXMH"
          data-testid="update-avatar-button"
          style="margin-top: 8px; margin-bottom: 16px; width: 320px;"
          type="button"
        >
          Update Avatar
        </button>
      </form>
      <div
        style="margin-bottom: 8px; margin-top: 32px;"
      >
        <div
          class="sc-gueXAH gRHGwY"
        >
          Full name
        </div>
        <input
          aria-label="Full name"
          class="sc-iAEzLF gTKcTi"
          placeholder="Enter your full name..."
          style="min-height: 48px; min-width: 320px; border: 1px solid #858585; background-color: rgb(254, 254, 255);"
          type="text"
          value="displayName"
        />
      </div>
      <button
        buttontype="primary"
        class="sc-gLDymS dItXMH"
        data-testid="update-button"
        style="margin-top: 8px; margin-bottom: 16px; width: 320px;"
        type="button"
      >
        Update
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
